webrtc/README.md

```diff
@@ -0,0 +1,130 @@
+# WebRTC
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active |                 |
+
+Authors: [@mxinden]
+
+Interest Group: [@marten-seemann]
+
+## Motivation
+
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Protocol
+
+Scenario: Browser _A_ wants to connect to server node _B_.
+
+### Setup - Both _A_ and _B_
+
+1. [Generate a
+   certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate).
+
+2. [Get the certificate's
+   fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+3. Sign this fingerprint using the libp2p peer key to generate a "peer cert"
+
+### Connection Establishment
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/<tls-certificate-fingerprint/p2p/<peer-id>`),
+  through some external mechanism.
+
+2. _A_ create a connection with its local certificate.
+
+3. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+4. _A_ establishes the connection to _B_.
+
+5. _A_ opens a data stream.
+
+6. _A_ and _B_ exchange the "peer certs".
+
+7. [Get the remote peer's TLS
+   certificate](https://www.w3.org/TR/webrtc/#dom-rtcdtlstransport-getremotecertificates).
+
+8. Check the remote endpoint's "peer cert".
```

Comment:
I want to point out that, while you can obtain the `RTCCertificate` object of the remote's certificate, but you can do basically nothing with it except check its expiration date ([see docs](https://developer.mozilla.org/en-US/docs/Web/API/RTCCertificate`)). It is not possible to read anything from the remote's certificate from a browser. The only thing you can know is that its fingerprint is the one that was in the SDP packet.
Step 7 is therefore useless, and we need to sign the fingerprint of the certificate with the peerid rather than the certificate itself.